Add property constraints-based matching support

dongahn opened the pull request:
This is a WIP PR that adds  property constraints-based matching support.

This builds on a pending PR #921 so don't merge this before landing that  PR.

- Add property support into flux ion-R encode: e.g., `flux R encode -r 0 -c 0-1 -g 0-1 -p "arm-v9@core:0" -H foo2 | flux ion-R encode` is now supported.
- Add properties support into libjobspec in accordance with new property addition to RFC14.
- Add property support within our new rv1exec reader
- Extend our job meta class to capture the property information
- Extend our traverser to use the properties captured in our job meta object to do perform property-based matching
- A bunch of test cases

I want to see if we can land this as part of this month's release? I still need to add a one end-to-end test case using `sched-fluxion-resource`.

grondo commented:
@dongahn, I realize this is still a WIP, but I gave this a try against current flux-core on fluke. I ran into some trouble so I thought I'd post my procedure (perhaps I'm doing something wrong).

To make it easy to launch a test instance from builddirs of flux-core and flux-sched, I made this little launcher script:

```sh
#!/bin/bash

declare -r long_opts="flux-core-dir:,flux-sched-dir:,properties:,nnodes:"
declare -r short_opts="f:F:p:N:"

eval set -- $(/usr/bin/getopt -u -o $short_opts -l $long_opts -n launcher -- $@)

FLUX=$(which flux)
while true; do
    case "$1" in
      -F|--flux-sched-dir) FLUX_SCHED_DIR=$2;     shift 2;;
      -f|--flux-core-dir)  FLUX=$2/src/cmd/flux;  shift 2;;
      -p|--properties)     PROPERTIES=$2;         shift 2;;
      -N|--nnodes)         NNODES=$2;             shift 2;;
      --)                  shift; break;                 ;;
      *)                   echo "Invalid option '$1'"; exit 1;;
    esac
done

test $# -eq 0 && set $SHELL

echo "Args: $@"

MODULE_PATH=${FLUX_SCHED_DIR}/resource/module/.libs
MODULE_PATH=${MODULE_PATH}:${FLUX_SCHED_DIR}/qmanager/module/.libs

echo "Launching instance with $FLUX on ${NNODES} nodes"
flux mini run -N ${NNODES} -n ${NNODES} \
    -o per-resource.type=node \
    -o pty.interactive \
    --env=FLUX_MODULE_PATH_PREPEND=${MODULE_PATH} \
    --env=PROPERTIES=$PROPERTIES \
    --env=ARGS="$@" \
    ${FLUX} start \
    bash -c '\
    echo setting properties to $PROPERTIES && \
    flux module remove sched-simple && \
    flux module remove resource && \
    R="$(flux kvs get resource.R | jq del\(.scheduling\))" &&
    flux kvs put resource.R="$(echo $R | flux R set-property $PROPERTIES)" && \
    echo establishing new resource.R && \
    flux kvs get resource.R | jq -S . && \
    echo loading core resource and Fluxion modules && \
    flux module load resource && \
    flux module load sched-fluxion-resource && \
    flux module load sched-fluxion-qmanager && \
    $ARGS'

# vi: ts=4 sw=4 expandtab
```

This was a quick and dirty script, so its usage is a little awkward, but basically specify the path to the flux-core and flux-sched directories to use, and the number of nodes to allocate, along with  a set of properties to apply and the script will launch a instance, set `resource.R` and load the fluxion modules, e.g.:

```console
$  ./test-launcher.sh -f . -F ~/git/flux-sched.git -p foo:0-1 -N2
Args: /bin/bash
Launching instance with ./src/cmd/flux on 2 nodes
setting properties to foo:0-1
establishing new resource.R
{
  "execution": {
    "R_lite": [
      {
        "children": {
          "core": "0-3"
        },
        "rank": "0-1"
      }
    ],
    "expiration": 1648865717,
    "nodelist": [
      "fluke[7-8]"
    ],
    "properties": {
      "foo": "0-1"
    },
    "starttime": 1648260917
  },
  "version": 1
}
loading core resource and Fluxion modules
flux-module: /usr/lib64/flux/modules/kvs.so: undefined symbol: flux_content_load
ƒ(s=2,d=1,builddir) grondo@fluke7:~/git/flux-core.git$ 
```

However, when I do this, `sched-fluxion-resource` is stuck, as evidenced by `flux resource list` hanging. GDB shows that the module is in a blocking RPC get:
```
Thread 18 (Thread 0x7fffaef52700 (LWP 1400056)):
#0  0x00007ffff65810f7 in epoll_wait () from /lib64/libc.so.6
#1  0x00007ffff7b95643 in epoll_poll (loop=0x7fff9c009370, 
    timeout=<optimized out>) at ev_epoll.c:155
#2  0x00007ffff7b97c3e in ev_run (flags=0, loop=0x7fff9c009370) at ev.c:3742
#3  ev_run (loop=0x7fff9c009370, flags=0) at ev.c:3623
#4  0x00007ffff7b67a1b in flux_reactor_run (r=0x7fff9c009350, 
    flags=flags@entry=0) at reactor.c:128
#5  0x00007ffff7b706e5 in flux_future_wait_for (f=f@entry=0x7fff9c009220, 
    timeout=timeout@entry=-1) at future.c:476
#6  0x00007ffff7b70839 in flux_future_get (f=f@entry=0x7fff9c009220, 
    result=result@entry=0x7fffaef513b0) at future.c:509
#7  0x00007ffff7b6da99 in flux_rpc_get_vunpack (ap=0x7fffaef513b8, 
    fmt=0x7fffaf36c11c "{s:o}", f=0x7fff9c009220) at rpc.c:172
#8  flux_rpc_get_unpack (f=f@entry=0x7fff9c009220, 
    fmt=fmt@entry=0x7fffaf36c11c "{s:o}") at rpc.c:172
#9  0x00007fffaf2d5d74 in grow_resource_db_hwloc (resobj=0x7fff9c008ab0, 
    ids=0x7fff9c008fb0, 
    ctx=std::shared_ptr<resource_ctx_t> (use count 2, weak count 0) = {...})
    at resource_match.cpp:825
#10 grow_resource_db (resources=<optimized out>, 
    ctx=std::shared_ptr<resource_ctx_t> (use count 2, weak count 0) = {...})
    at resource_match.cpp:1017
#11 update_resource_db (down=<optimized out>, up=<optimized out>, 
    resources=<optimized out>, 
    ctx=std::shared_ptr<resource_ctx_t> (use count 2, weak count 0) = {...})
    at resource_match.cpp:1131
#12 update_resource (f=0x7fff9c007580, arg=<optimized out>)
    at resource_match.cpp:1164
#13 0x00007fffaf2d6ec4 in populate_resource_db_acquire (
    ctx=std::shared_ptr<resource_ctx_t> (use count 2, weak count 0) = {...})
    at /usr/include/c++/8/bits/shared_ptr_base.h:1018
#14 populate_resource_db (
    ctx=std::shared_ptr<resource_ctx_t> (use count 2, weak count 0) = {...})
    at resource_match.cpp:1224
#15 init_resource_graph (
    ctx=std::shared_ptr<resource_ctx_t> (use count 2, weak count 0) = {...})
    at resource_match.cpp:1322
#16 0x00007fffaf2d7b25 in mod_main (h=0x7fff9c001b70, argc=<optimized out>, 
    argv=0x7fff9c01a3e0) at resource_match.cpp:2527
#17 0x000000000040d927 in module_thread (arg=0x6931a0) at module.c:183
#18 0x00007ffff772017a in start_thread () from /lib64/libpthread.so.0
--Type <RET> for more, q to quit, c to continue without paging--
#19 0x00007ffff6580dc3 in clone () from /lib64/libc.so.6
```

grondo commented:
It occurs to me that I built flux-sched in this case against the installed flux-core, instead of my working copy with properties support. I'm not sure if that is why I ran into trouble here.

dongahn commented:
>It occurs to me that I built flux-sched in this case against the installed flux-core, instead of my working copy with properties support. I'm not sure if that is why I ran into trouble here.

I'm not sure either. But two things are suspicious.

> flux-module: /usr/lib64/flux/modules/kvs.so: undefined symbol: flux_content_load
Is this still safe? 

>#9  0x00007fffaf2d5d74 in grow_resource_db_hwloc (resobj=0x7fff9c008ab0, 
    ids=0x7fff9c008fb0, 

I don't think this should be called. `grow_resource_db_rv1exec` should be called instead. So one possibility is the rv1 object passed from `resource.acquire` somehow cause flux-resource to call the incorrect reader.

Let me look a bit more.

grondo commented:
> flux-module: /usr/lib64/flux/modules/kvs.so: undefined symbol: flux_content_load

Sorry, I should have mentioned this was an expected error. This is because I'm running flux-core out of my build directory, and that symbol was removed from the exported interface recently. But `flux module` still scans the entire library path for potential modules to get their mod_name, and this path includes the system library paths. The test instance obviously doesn't actually load the system kvs module which includes this symbol.

dongahn commented:
>#10 grow_resource_db (resources=<optimized out>, 
    ctx=std::shared_ptr<resource_ctx_t> (use count 2, weak count 0) = {...})
    at resource_match.cpp:1017

@grondo: line 1017 matches with the code at the main branch... Could it be the case that a different flux-sched in the system is being loaded?

grondo commented:
Yes, that is a possibility! Sorry for the confusion. Let me add a way to test that the expected modules are loaded and get back to you.

grondo commented:
I guess the open question then is why the system fluxion modules are getting stuck?

dongahn commented:
>I guess the open question then is why the system fluxion modules are getting stuck?

It is stuck at https://github.com/flux-framework/flux-sched/blob/master/resource/modules/resource_match.cpp#L825

>resource.get-xml

Is this blocking until the hwloc xml is gathered?

Is it possible 

>flux module load resource

with the new R, doesn't gather xml?

grondo commented:
> Is this blocking until the hwloc xml is gathered?

Yes, that was the problem. In my test I was only reloading the `resource` module on rank 0, whereas it needs to be reloaded across all ranks or the xml cannot be re-populated, and the `get-xml` RPC will hang.

grondo commented:
Ok, I now have a fixed test script:

```sh

```

Within the test instance `flux mini` commands work with `-n`, but when I try `-N2 -n2` I get the following error:

```
ƒ(s=4,d=1,builddir) grondo@fluke7:~/git/flux-core.git$ flux mini run -N2 -n2 hostname
2022-03-29T19:52:47.131778Z sched-fluxion-resource.err[0]: match_multi_request_cb: match failed due to match error (id=502930604032): Success
2022-03-29T19:52:47.131810Z sched-fluxion-resource.err[0]: match_multi_request_cb: flux_respond_error: Invalid argument
^Cflux-job: one more ctrl-C within 2s to cancel or ctrl-Z to detach
^Zflux-job: detaching...
```

```
ƒ(s=4,d=1,builddir) grondo@fluke7:~/git/flux-core.git$ flux job info ƒEDFFoju jobspec | jq .resources
[
  {
    "type": "node",
    "count": 2,
    "with": [
      {
        "type": "slot",
        "count": 1,
        "with": [
          {
            "type": "core",
            "count": 1
          }
        ],
        "label": "task"
      }
    ]
  }
]
```

dongahn commented:
> Within the test instance flux mini commands work with -n, but when I try -N2 -n2 I get the following error:

@grondo: I am still struggling with another issue I am seeing. But in the meantime, your script might still use the installed fluxion. Can you try the following and see if you still see the error?

```diff
22,23c22,25
< MODULE_PATH=${FLUX_SCHED_DIR}/resource/module/.libs
< MODULE_PATH=${MODULE_PATH}:${FLUX_SCHED_DIR}/qmanager/module/.libs
---
> MODULE_PATH=${FLUX_SCHED_DIR}/resource/modules/.libs
> MODULE_PATH=${MODULE_PATH}:${FLUX_SCHED_DIR}/qmanager/modules/.libs
>
> echo $MODULE_PATH
34a37
>     echo ${FLUX_MODULE_PATH_PREPEND} && \
```

grondo commented:
Yeah, sorry! I found that typo yesterday but forgot to update the script I pasted here. My apologies.

I will try this branch again with the fixed script today.

grondo commented:
@dongahn, I tried out this branch again (with the fix for the launcher script) and a basic request does seem to be working:
```console
$ flux kvs get resource.R | jq .execution.properties
{
  "foo": "1-3"
}
$ flux mini run -N3 --requires=foo flux getattr rank
2
3
1
```

However, the `^` (not) character generates an error and a job hang:
```console
$ flux mini run -N3 --requires=^foo flux getattr rank
2022-04-01T16:39:58.252590Z sched-fluxion-qmanager.err[0]: jobmanager_alloc_cb: ENOSYS: ^foo is invalid: Function not implemented
```

Also, the properties don't seem to be propagated into the R for jobs, nor in the R sets returned from the `sched.resource-status` RPC:

```console
$ flux job info $(flux mini submit -N3 --wait-event=start --requires=foo flux getattr rank) R | jq
{
  "version": 1,
  "execution": {
    "R_lite": [
      {
        "rank": "1-3",
        "children": {
          "core": "3"
        }
      }
    ],
    "nodelist": [
      "fluke[8-10]"
    ],
    "starttime": 1648831437,
    "expiration": 1649436237
  }
}
$ flux resource list -no {properties}
$
```

dongahn commented:
@grondon: oh great! Yes the exporting properties is still WIP. I don't have any support for operators other than simple "properties" since I just wanted to just a minimum viable within my limited time budge. Should additional operators be considered as part of this PR?

dongahn commented:
>However, the ^ (not) character generates an error and a job hang:

And yes, no matter what the hang shouldn't happen. Great testing!

grondo commented:
The exclusion character `^` is required by the current version of [RFC 31](https://flux-framework.readthedocs.io/projects/flux-rfc/en/latest/spec_31.html):

> As a special case, if a property value begins with the character ^, then the remaining part of the value SHALL indicate a property that MUST NOT be included in the allocated resource set.

If that is too difficult to implement at this time we can relax that requirement temporarily.

However, either way, there is a problem because a job with an invalid property seems to get stuck in an unkillable RUN state.

dongahn commented:
>@grondo: I am still struggling with another issue I am seeing. 

I posted a PR (#924) for this problem -- without this fix I had to pull my hair out testing this capability on a single node docker platform.

dongahn commented:
Still want to finish this, but other priorities (and vac today) are preventing at the moment. I will see if I can spend a bit more time once I need to make progress I need to make to deal with the LDRD priorities. Thanks @grondo for your early comments, testing and reviews!!

dongahn commented:
>> However, the ^ (not) character generates an error and a job hang:

> And yes, no matter what the hang shouldn't happen. Great testing!

This was essentially due to not catching the exception from `libjobspec` on the spot. Easy to fix. Also handling `^` turns out to be pretty easy to add as well.

dongahn commented:
The only major thing that still remains to be done is property propagation into R. Will take a look at it tomorrow.

dongahn commented:
>The only major thing that still remains to be done is property propagation into R. Will take a look at it tomorrow.

@grondo: Do you export properties into a job's `R` with `sched-simple`?

grondo commented:
> @grondo: Do you export properties into a job's R with sched-simple?

Yes:

```
ƒ(s=8,d=0,builddir) grondo@asp:~/git/flux-core.git$ flux kvs put resource.R="$(flux kvs get resource.R  | flux R  set-property xx:0-2 yy:2-7)"
ƒ(s=8,d=0,builddir) grondo@asp:~/git/flux-core.git$ flux module load resource
ƒ(s=8,d=0,builddir) grondo@asp:~/git/flux-core.git$ flux module load sched-simple
ƒ(s=8,d=0,builddir) grondo@asp:~/git/flux-core.git$ flux resource list
     STATE PROPERTIES NNODES   NCORES    NGPUS NODELIST
      free xx              2        8        0 asp,asp
      free xx,yy           1        4        0 asp
      free yy              5       20        0 asp,asp,asp,asp,asp
 allocated                 0        0        0 
      down                 0        0        0 
ƒ(s=8,d=0,builddir) grondo@asp:~/git/flux-core.git$ flux mini alloc -N1 flux resource list
     STATE PROPERTIES NNODES   NCORES    NGPUS NODELIST
      free xx              1        4        0 asp
 allocated                 0        0        0 
      down                 0        0        0 
ƒ(s=8,d=0,builddir) grondo@asp:~/git/flux-core.git$ flux mini alloc -N2 --requires yy --requires ^xx sh -c 'flux kvs get resource.R | jq'
{
  "version": 1,
  "execution": {
    "R_lite": [
      {
        "rank": "0-1",
        "children": {
          "core": "0-3"
        }
      }
    ],
    "starttime": 0,
    "expiration": 0,
    "nodelist": [
      "asp,asp"
    ],
    "properties": {
      "yy": "0-1"
    }
  }
}
```

dongahn commented:
Thanks!  Just one more clarification. Not only matching properties but also other properties should be exported, correct? (IOW, all of the properties of each and every allocated rank.)

grondo commented:
Yes, that is correct

On Fri, May 13, 2022, 8:51 AM Dong H. Ahn ***@***.***> wrote:

> Thanks! Just one more clarification. Not only matching properties but also
> other properties should be exported, correct? (IOW, all of the properties
> of each and every allocated rank.)
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/flux-framework/flux-sched/pull/922#issuecomment-1126199667>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAFVEUTRZA7OJ4JPMQAOV23VJZ3BFANCNFSM5RS2FRFA>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

dongahn commented:
@jameshcorbett and @grondo: this PR passes all of the CI tests now and ready for your review, I think!

jameshcorbett reviewed (approved):
I looked through the code (but didn't run any tests) and things look reasonable. One question below.

jameshcorbett commented on resource/readers/resource_reader_rv1exec.cpp:
```diff
@@ -238,12 +285,69 @@ int resource_reader_rv1exec_t::build_rmap (json_t *rlite,
+            // the following code can be extended: fetching @resource_type and
+            // add the property to the corresponding resource_type vertices.
+            // For now, we just add the property to compute node vertices.
+            if (pmap.find (rank) != pmap.end ()) {
```
How does this interact with something like ``--requires="arm-v9@core"`` which was in the tests?

dongahn replied:
>So right now only nodes can have properties?

Yes that's the current level of support with RV1 RFC https://github.com/flux-framework/rfc/blob/master/spec_20.rst

dongahn replied:
>How does this interact with something like --requires="arm-v9@core" which was in the tests?

The property (`arm-v9@core`) as a whole will be stored in the corresponding compute node vertices and properties will be matched there. We wanted to have scheduler specific hints directly in the properties like `@core` in this case for more advanced (fined-grained) property-based constraints. (Also documented at https://github.com/flux-framework/rfc/blob/master/spec_20.rst).

Hope this helps.

grondo commented:
I haven't looked at this yet, but I've just started testing. One question: are the properties transmitted in a `sched.resource-status` response? This would be required in order for `flux resource list` to be able to list properties. I'm trying to see if I've got my test setup correctly.

dongahn commented:
>I haven't looked at this yet, but I've just started testing. One question: are the properties transmitted in a sched.resource-status response?

No they are not! Didn't realize that was augmented. Will do a few quick commits and tests for that. If you can add the code in flux-core or examples on this augmented format, this will help me finishing this up quicker. Thanks @grondo!

grondo commented:
> . If you can add the code in flux-core or examples on this augmented format, this will help me finishing this up quicker.

Sorry, perhaps I wasn't clear. When you emit an Rv1 from any resource set, it should include any properties currently set on that resource fragment. Since the `sched.resource-status` response includes Rv1 fragments, these should all include respective properties.

I actually made the following addition to one of your existing tests and it looks like you _do_ already have support for this:

```diff
diff --git a/t/t1022-property-constraints.t b/t/t1022-property-constraints.t
index f0a78f28..9d97e895 100755
--- a/t/t1022-property-constraints.t
+++ b/t/t1022-property-constraints.t
@@ -21,6 +21,11 @@ test_expect_success 'load fluxion' '
        load_qmanager
 '
 
+test_expect_success 'properties are listed in flux resource list' '
+       flux resource list &&
+       flux resource list | grep foo
+'
+
 test_expect_success 'simple property reqirement works' '
        JOBID1=$(flux mini submit --requires="baz" hostname) &&
        flux job wait-event -t 10 ${JOBID1} clean &&
```

``expecting success: 
	flux resource list &&
	flux resource list | grep foo

     STATE PROPERTIES NNODES   NCORES    NGPUS NODELIST
      free foo             1        4        0 fluke108
      free foo,bar         1        4        0 fluke108
      free bar             1        4        0 fluke108
      free baz             1        4        0 fluke108
 allocated                 0        0        0 
      down                 0        0        0 
      free foo             1        4        0 fluke108
      free foo,bar         1        4        0 fluke108
ok 3 - properties are listed in flux resource list`
```

So I must be doing something incorrect in my manual testing.

dongahn commented:
Ah .. good. Yes, I made the changed in rv1 and rv1_nosched writer so anything that uses rv1 objects should have property support. Sorry for the confusion and thanks!

dongahn commented:
>So I must be doing something incorrect in my manual testing.

There could still be a bug so keep your feedback coming!

grondo commented:
My guess is I was using the system-installed fluxion modules inadvertently, thanks! :facepalm:

grondo commented on t/t1022-property-constraints.t:
```diff
@@ -0,0 +1,76 @@
+'
+
+test_expect_success 'removing resource and qmanager modules' '
+    remove_qmanager &&
+    remove_resource
```
whitespace error here, should be tabs like the rest of the file.

grondo reviewed (approved):
I ran through some testing on fluke and everything seemed to work as expected. I only found one very minor whitespace issue (noted inline).

It might be useful to leave some notes or instructions on how one might go about supporting some of the optional RFC 31 constraint operators in the future such as `or` and `not`. That might come in handy in the near term, but after a cursory look through the diff here, I would have no idea how to go about it.

One of our todo items is to allow exclusion or inclusion of hosts in a list and it seems like we could add a couple new operators to our constraint specification to easily satisfy this use case.

dongahn commented:
>I ran through some testing on fluke and everything seemed to work as expected. I only found one very minor whitespace issue (noted inline).

Thanks!

>It might be useful to leave some notes or instructions on how one might go about supporting some of the optional RFC 31 constraint operators in the future such as or and not. That might come in handy in the near term, but after a cursory look through the diff here, I would have no idea how to go about it.

> One of our todo items is to allow exclusion or inclusion of hosts in a list and it seems like we could add a couple new operators to our constraint specification to easily satisfy this use case.

Yes, I will do that. 

Also for using properties for partitions, maybe we can open up a ticket and I can leave some thought there as to how we can support that as well.

dongahn commented:
Thanks @grondo and @jameshcorbett! I addressed the whitespace issue and forced a push before adding MWP. 

I also added a note at https://github.com/flux-framework/flux-sched/issues/942 with code change suggestions to support fuller RFC 31 property constraints specification support.
